tsl/src/fdw/data_node_scan_plan.c

```diff
@@ -674,6 +688,689 @@ push_down_group_bys(PlannerInfo *root, RelOptInfo *hyper_rel, Hyperspace *hs,
+
+	return true;
+}
+
+/*
+ * Check if the join fromList belongs to an implicit join (e.g.,
+ * SELECT * FROM metric m1, metric_name m2 WHERE m1.id=m2.id). This type
+ * of join has two RANGETBLREF entries in the fromList.
+ */
+static bool
+is_implicit_join_two_tables(List *fromList)
+{
+	if (list_length(fromList) != 2)
+		return false;
+
+	/* At the moment, the code is called only in the reference table join
+	 * pushdown path. Earlier checks ensure that both parameters are
+	 * RangeTblRefs. Recheck the condition here and handle the case properly.
+	 */
+	if (!IsA(list_nth(fromList, 0), RangeTblRef))
+		return false;
+
+	if (!IsA(list_nth(fromList, 1), RangeTblRef))
+		return false;
+
+	return true;
+}
+
+/*
+ * Check if the query performs a join between a hypertble (outer) and a reference
+ * table (inner) and the join type is a LEFT JOIN, an INNER JOIN, or an implicit
+ * join.
+ */
+static bool
+is_safe_to_pushdown_reftable_join(PlannerInfo *root, List *join_reference_tables,
+								  RangeTblEntry *outertableref, RangeTblEntry *innertableref,
+								  JoinType jointype)
+{
+	Query *query = root->parse;
+	List *fromList;
+
+	Assert(root != NULL);
+	Assert(outertableref != NULL);
+	Assert(innertableref != NULL);
+
+	/*
+	 * We only reach this code path if the outer table is a
+	 * distributed hypertable.
+	 */
+	Assert(ts_is_hypertable(outertableref->relid));
+
+	/*
+	 * We support pushing down of INNER and LEFT joins only.
+	 *
+	 * Constructing queries representing partitioned OUTER, SEMI and ANTI
+	 * joins is hard, hence not considered right now.
+	 */
+	if (jointype != JOIN_INNER && jointype != JOIN_LEFT)
+		return false;
+
+	/* Check that at least one reference table is defined. */
+	if (join_reference_tables == NIL)
+		return false;
+
+	/* Only queries with two tables and one join are supported. */
+	fromList = query->jointree->fromlist;
+	if (!is_inner_or_left_join_two_tables(fromList) && !is_implicit_join_two_tables(fromList))
+		return false;
+
+	/* Right table has to be a distributed hypertable */
+	if (!list_member_oid(join_reference_tables, innertableref->relid))
+		return false;
+
+	/* Join can be pushed down */
+	return true;
+}
+
+/*
+ * Assess whether the join between inner and outer relations can be pushed down
+ * to the foreign server. As a side effect, save information we obtain in this
+ * function to TsFdwRelInfo passed in.
+ */
+static void
+fdw_pushdown_foreign_join(PlannerInfo *root, RelOptInfo *joinrel, JoinType jointype,
+						  RelOptInfo *outerrel, RelOptInfo *innerrel, JoinPathExtraData *extra)
+{
+	TsFdwRelInfo *fpinfo;
+	TsFdwRelInfo *fpinfo_o;
+	TsFdwRelInfo *fpinfo_i;
+	ListCell *lc;
+	List *joinclauses;
+
+	/*
+	 * If either of the joining relations is marked as unsafe to pushdown, the
+	 * join can not be pushed down.
+	 */
+	fpinfo = fdw_relinfo_get(joinrel);
+	fpinfo_o = fdw_relinfo_get(outerrel);
+	fpinfo_i = fdw_relinfo_get(innerrel);
+
+	/* shouldn't happen in the current implementation */
```

Comment:
Consider asserting for now if it cannot happen. We can always change asserts to checks later, when needed.

Reply:
We are now returning (and not pushing down the join) instead of showing an error message. I could change this to an assert statement. But I am more in favor of returning this case. What is your opinion on this?

Reply:
I have converted the statements to `Asserts.`